### ISSUE TITLE
UI improvements

### ISSUE THREAD

**Issue by howardchung** [yasp-dota/yasp#410]
- [x] Sans-serif headings?  I think they're quite polarzing--some people seem to hate them a lot.
- [x] Remove fixed header?  Seems like most of the other Dota 2 sites do not use one.
- [x] cheese_block doesn't need its own special section on home page (can go in footer like on every other page)
- [ ] numeral rounding in match tables?  Only if we choose to add back HD/TD/HH
- [ ] Denote R/D in gold/xp/lh graphs?
- [ ] Better cheese icon?  (I think the current icon requires too much css/custom code to get it working, plus we have an extra line in the footer just for attribution)  Just use the Dota 2 cheese item image?
- [ ] Material design?  Requires adding class names to all our templates :(
- [ ] Use a sidebar for tab navigation on player/match pages

**Comment by howardc93** [comment]
I don't particularly like our font, either.  Perhaps it's fine for the YASP on the front page, but I'm not sure about using it for all the headings.

**Comment by howardc93** [comment]
Color contrast:  I think the dark theme should be darker (rather than the dark gray it is currently), and we should choose better colors for "success" and "danger".

**Comment by howardc93** [comment]
I've stated this a few times, but I still think it's unnecessary to have two color themes.  It requires an option UI, an API endpoint to make updates, and extra testing to make sure style changes are okay in both themes.  Very few sites, in my experience, have theme selection (the default theme should be acceptable to everyone)

**Comment by howardc93** [comment]
@albertcui has resisted removing the theme selector pretty strongly in the past, does anyone else have an opinion?

**Comment by mvthen** [comment]
I think there are some people who would appreciate the light theme - do your analytics say anything about the use of either?

**Comment by howardc93** [comment]
We can count the number of users who have it enabled.

I know @nickhh uses it.

**Comment by albertcui** [comment]
Yeah I was just gonna say we should just run a query.

**Comment by howardc93** [comment]
db.players.count({last_visited:{$ne:null}, light_theme:true})
717

out of 11440 all-time users

**Comment by howardc93** [comment]
db.players.count({last_visited:{$gt:new Date("2015-05-20")}, light_theme:true})
383

out of about 3000 users in the last week

**Comment by mvthen** [comment]
~13%. That’s not bad. Was about to say if it’s anything less than 20% maybe cut it, but as long as it’s not a huge add-on, don’t see why it can’t stay (though I get you’re trying to minimize).

**Comment by howardc93** [comment]
If we want to spice up bootstrap:

https://bootswatch.com/

**Comment by alexbarrett** [comment]
As a user of both YASP and Dotabuff I think there are currently 2 things stopping people from using switching to YASP as their primary Dota 2 stats tracker:

* It's slow as balls
* The UI sucks

Statswise it's far better than Dotabuff (at least for analyzing matches) but these two issues are really holding it back.

I can't really help with the first one as I'm not familiar enough with the code and infrastructure, but I can definitely help with the second! Anyway I've got loads of feedback and suggestions and I thought I'd splurt most of it out here.

#### yasp.co always goes to the landing page
This is great when you've never visited the site before, but once it's been read you don't really ever want to see it again.

**Solution:** Redirect logged in users to their profile page when hitting the site index.

#### Fixed Header
Fixed headers are a *really* prominent UI component so if you're gonna have one you better make damn sure it's useful. The fixed header currently contains top-level navigation items that make sense to browse to from the landing page (or another top-level page), but 99% of the time are going to be irrelevant to the user who's looking at their matches (which is what peopl are going to be doing 99% of the time on YASP).

What would be useful? The profile tabs (player, matches, histograms etc.)

**Solution:** Replace the fixed header navigation items on profile pages with what are currently displayed as tabs.

#### Colors
You've already identified that the green/red used are pretty garish and could be improved. A bigger issue (in my opinion) is that the same green and red are currently being used to indicate Radiant/Dire **as well as** good/bad (e.g. above/below 50% win rate). This adds an element of confusion when scanning the pages and I think this is the main reason Dotabuff chose not to colour things for Radiant/Dire as a general rule (they have some really faint shading but not enough to confuse things).

**Solution:** Use green/red for either Radiant/Dire or good/bad, not both.

#### Bootstrap grid

This is more of a technical issue. For some reason the negative left/right
margin has been removed from the Bootstrap `.row` class which is causing nested
rows to appear indented. This is particularly noticable on tables which all have
some extraneous horizontal padding due to this.

**Solution:** Resore the -15px margin-left/right on the `.row` class.

#### Table headers

Table headers wrap (both text content and sort indicators) which makes them look messy.

**Solution:** Set `white-space: nowrap` and remove the current neutral sort indicator (class="sorting") and just show asc/desc indicators where applicable.

#### Table pagination

Currently pagination is displayed as a combination of an "amount" above the table and a large prev/next control below. This is rather a lot and I think we could take a leaf out of Google's snazzy new [material design spec](http://www.google.co.uk/design/spec/components/data-tables.html#data-tables-tables-within-cards) and replace it with something more compact.

#### Table content

Currently some of the data in tables is pretty weird or could be improved:

##### Matches

* **Match ID** - This is the widest table cell and 90% of the time users won't care about it. Recommend moving this somewhere else (not in table listings).
* **Status** - Whether or not a match has been parsed doesn't deserve a large column like this. Recommend removing it or indicating it some other, subtler way (a plus icon or something).
* **Game mode** - I think ranked should be a separate field as it is indepentantly useful for searching/filtering. Bonus points if you can differenciate solo/party (I know that would be difficult).
* **Numbers** - Dotabuff rounds large numbers which I think is a lot more readable and just as useful most of the time (e.g. 16312 -> 16.3k). The exact value could be shown as a tooltip or something.
* **Win/Loss** - Only indicated by color not in the table itself. Dotabuff displays this much more nicely.

##### Teammates

Matches/Teammate/Opponent columns. For most people this is going to be people they party with so opponent value will be zero. It's also called "teammates" so why's it showing opposition anyway.

Think what you could show here instead! How about a nice little head-to-head comparing your performance vs your friend's performance in your recent games together? Who's carrying who? A generalised head-to-head feature that operates over a set of games (and integrates with filters) would make really nice feature on it's own actually.

*This is turning out longer than I antipated. I'll be more brief from now on...*

#### Tabs

* **Player tab** - Should be renamed to overview (all tabs are player tabs).
* **Advanced tab** - Should be renamed or the content reappropriated into other tabs.
* **Steam profile** - This should not be a tab, it's an offsite link. Make it a nice little Steam icon next to the name or something.

#### Player/Overview

Think about what people want to see most when seeing an overview of their own games or other peoples' games.

#### Cheese icon

Ugly. Can I donate to remove it?

#### Profile header

Full history, name update and last visit. LMFAO wHo cAREAS HAHA Xd

Look at what Dotabuff puts here, it's *so* much better.

#### Graphs

Colors could be better (I like the suggestion of using the in game colors I saw in a different issue).

Radiant/Dire heroes should be separated in the legend.

The hover feature is nice but I'd also like to be able to toggle lines on/off entirely.

#### Filters

Personally I think this is YASP's best feature over Dotabuff. Every tab is essentially a view over a set of matches, and any view's set of matches can be filtered in any way. This gives much more flexibility for viewing data than Dotabuff offers.

Let's make it better! Add more filters. Improve the filtering interface. Make the active filters follow you across each tab. This is YASP's killer feature.

### New features

#### Sparklines

Lots of opportunities to use sparklines in various tables.

#### Trends & Records

These are my favourite tabs on Dotabuff. Imitation is the best form of flattery. I imagine you'll tell me this stuff already exists in various places (histograms or places within various tabs), but it's not quite the same.

#### Friends

Let people mark other users as friends. To get started these players can just be added to the player dropdown filter, but it can also be used for much more.

#### In reference to earlier suggestions

> Smaller text

Small text should be avoided in my opinion. The current base font size of 14px is small enough as it is. I can see why it might be needed sometimes as YASP packs quite a lot of information on the screen, but it should be used sparingly.

Maybe you're comparing YASP to Dotabuff and seeing that they have a 12px base font size, but small text is not the reason Dotabuff looks better than YASP.

> I don't particularly like our heading font, either. Perhaps it's fine for the YASP on the front page, but I'm not sure about using it for all the headings.

I think the font is fine but most headings are too big and misaligned (relative to their content) which makes the font look bad.

> Color contrast: I think the dark theme should be darker (rather than the dark gray it is currently), and we should choose better colors for "success" and "danger".

It's pretty much the same as Dotabuff, I don't see a problem with the shade of gray used. It could be changed but it's not an issue at all.

> I still think it's unnecessary to have two color themes

100% agreed. The UI could use a lot of improvement and having one theme will greatly simplify that task.

----

I could keep going but this is already quite a lot. Some of these should be broken into separate issues but I didn't want to litter the issue tracker and this issue seems like a decent place for discussion. I'll try and contribute some things myself if I can fit it around my day job (and Dota addiction). In the meantime, what do you all think?

**Comment by howardc93** [comment]
Thanks so much for the feedback!  I agree with you on most of these points, and we'll work to implement those changes.

Slowness:
We're aware there's currently quite a bit of slowdown.  It's particularly bad when we're processing a lot of full histories (such as right now, since we added about 600 new users yesterday).  We recently added a cache to make player index page loads faster, but every other tab still requires iteration through all of a player's matches, which can be really slow.  We may need to look into upgrading our hardware...

landing page:
We used to have an "dashboard" for logged in users, and I think @albertcui agrees with you.  I swapped it back mainly for easier debugging purposes, but I don't object to changing it back.  Redirect to profile sounds reasonable to me.

Fixed header:
Agree.  I never really liked the fixed header anyway... and it's consuming valuable screen real estate.

Colors:
It seems like the solution is to either use a different set of shades for good/bad and radiant/dire, or switch to text for one of them (which I think is what Dotabuff is doing).  Which one is better?

Bootstrap grid:
@albertcui will have to look into this, I don't recall changing this.  This is part of a reason I want to keep our custom css minimal.

headers/pagination:
These are generated through the external library (DataTables) that we're using.  We can disable the "matches per page" if it's not very useful.

Match content:
Mostly agree.

Teammates:
We used to only count players you were in games with.  I recently added opponents as well (since some people play the same people repeatedly).  Since I was adding it anyway it was trivial to display with/against for each player.  I realize it's not very applicable for the "top" players, but not sure how to display this cleanly in the UI...

Comparisons: check out https://github.com/yasp-dota/yasp/issues/351!
We will need to work on building a nice UI, but the infrastructure is there.

Tabs:
Agree, but need to figure out how to organize the data in Advanced.  I suppose we can break it down into build times, accuracy, wardmaps, killstreaks/multikills.
@albertcui can take care of the steam profile thing?  I agree it doesn't deserve its own tab.

Overview:
This currently has most recent 10 matches, all-time W/L, heroes played, teammates/opponents, and rating history.  Anything else?

Cheese icon:
Will need to see if @albertcui agrees.  I am neutral on the icon, but I do dislike the amount of custom code we had to commit just to get it working.

Header:
TBH, most of this was just debugging information for us.  Perhaps we should just make it less prominent...

Graphs:
We'll see what we can do.  Separating the legend and disabling lines may require playing with c3...  Colors should be easy and added soon @albertcui ?

Filters:
Glad you like it.  I am not sure how to improve beyond just a set of select/dropdown menus, though.  What other filters would you like to see?
I've thought about trying to preserve the query string.  We'll need a nice clean way to "reset" the query though, which can be done right now by clicking any tab.

Sparklines:
Perhaps http://benpickles.github.io/peity/
I'm not sure how much useful information they add, but they look nice!

Trends/records:
I'm open to someone submitting work to implement these, but I'm not particularly interested in doing it myself.  Trends require some kind of bucketing in order to not have wild variance between matches.  Records already exist on histograms page but I suppose could be moved to its own tab.

Friends:
In the latest dev code I populate the account_id dropdown with a player's teammates, so that's somewhat resolved.  Perhaps one day we'll add the ability for a user to add "favorite" users to a list.

Misalignment I think is usually due to inconsistency in placing the heading within a row/col set.  We should just find these and clean them up.

By all means, keep going!  I found this comment extremely helpful :)

We might make separate issues once we actually start implementing things.

**Comment by howardc93** [comment]
More notes regarding speed:

I think the critical issue to make the common case fast.  This I think is, for a typical session:

* User plays match
* User visits YASP player page (bookmarked or browser history)
* User clicks link to their most recent match
* User browses match pages

So the cache should make the index player page load fast (the cache is missed if they visit a non-index tab or if they submitted a query).
Match pages have always been pretty quick since they don't require iterating over multiple matches and joining/aggregating data.

**Comment by howardc93** [comment]
Switching to SSD might also improve load times, at considerably more expense...

**Comment by howardc93** [comment]
Note on trends/records:

Dotabuff does a pretty good job of these already.  We would only be able to improve on that by taking advantage of our parsed data for new trends/records categories.

We don't intend to replace Dotabuff completely.  I still use it myself, mainly for checking global hero winrates immediately after a patch, or seeing global hero matchup percentages.  The goal is to have YASP be robust enough for "everyday" tasks such as checking stats of the latest matches you play (and getting free replay parses on them).  Anything else is really icing on the cake.

**Comment by howardc93** [comment]
Could we do a better job than Dotabuff overall?  Maybe, but to do so we'd need to invest a lot of money to get the capacity required to handle the full Dota 2 match history.

Dotabuff has smart, motivated full-time employees and they do an excellent job.  They also have very significant market share and brand recognition, and so I prefer to work on features that they don't already offer for free.

**Comment by alexbarrett** [comment]
> Bootstrap grid:
> I don't recall changing this. This is part of a reason I want to keep our custom css minimal.

I did a bit of digging and found the problem in [d83f896](https://github.com/yasp-dota/yasp/commit/d83f896471a0b92d1d9fe8f84a2aaf285eaa93f8#diff-d584a6010c08d4019c8b34945ee3175fR9). Those styles on `.row` shouldn't be overridden globally so if there is a scrolling bug it needs to be fixed another way. Once this line is removed the layout will align much better.

> headers/pagination:
> These are generated through the external library (DataTables) that we're using. We can disable the "matches per page" if it's not very useful.

I like it, but it should be more compact. Also it and all other pagination UI should be hidden for tables with up to 10 results (e.g. player's recent matches).

> Teammates: ...
> Comparisons: ...
> Tabs: ...
> Overview: ...

In general all these things (and others) need to be given some thought about what people expect to see and when/how they see it. Take the heroes table on the player profile as just one example:

When I look at my own (or someone else's) profile I expect to see my most played heroes and win rates with those heroes. I don't expect to see the number of times they've played *against* their most played heroes. It's not that the data isn't useful, it's just in a strange place.

> Sparklines:
> Perhaps http://benpickles.github.io/peity/
> I'm not sure how much useful information they add, but they look nice!

An obvious candidate would be match GPM/XPM but that may not be that useful. When I mentioned sparklines though I was mostly thinking about a feature I haven't suggested yet...

If they ever are added it should be possible to implement them with d3.js as that's already a dependency.

> We don't intend to replace Dotabuff completely

I guess it depends on the project's goals. The homepage says "free replay parsing for all" which is pretty open-ended. Personally I would just like to make the best replay analysis tool possible for improving your game and seeing data in interesting ways. Dotabuff is just the most obvious point of comparison.

> Could we do a better job than Dotabuff overall? Maybe, but to do so we'd need to invest a lot of money to get the capacity required to handle the full Dota 2 match history.

Dotabuff really is amazing. While replying here I've been looking at Dotabuff's interface a lot and noticing just how good the UX really is. Everything is extremely well thought out.

Making YASP as good as possible for the "everyday" tasks as you say seems like a sensible direction, at least for the time being.

> We may need to look into upgrading our hardware...

What is YASP currently running on?

**Comment by howardc93** [comment]
margin problem fixed in ui branch.

We're running on GCE, so it's just a matter of selecting a larger instance.  It'll cost more money, so we need to budget carefully to make sure the costs don't outstrip our donations by too much.

**Comment by howardc93** [comment]
On the topic of displaying with/against data:

YASP is primarily intended as a personal analysis tool.  While you CAN view other player profiles, there's no guarantee that we have all their matches (and most of the time, we don't).  So the expected use case is for players to view their own profiles, and for that it seems logical to display data such as with/against for each hero they commonly play.

**Comment by alexbarrett** [comment]
>> **Status** - Whether or not a match has been parsed doesn't deserve a large column like this. Recommend removing it or indicating it some other, subtler way (a plus icon or something).
> I think it's fine as is. It's not taking up THAT much space.

Space aside, what is the reason for keeping it? I don't understand when it would ever be useful. As a filter it makes sense, but why would anyone ever want to compare parsed status between rows or sort on this value?

> It IS possible to determine ranked/unranked (ranked games have a different lobby_type, but this is a column we aren't currently displaying. We'd need to add another column in addition to game_mode. I don't know if it's worth the additional space.

Ah yes, I can see now in the filters that there are separate fields. I see you already differentiate between solo/party as well. That's really nice. (How did you do it?)

>> Numbers - Dotabuff rounds large numbers which I think is a lot more readable and just as useful most of the time (e.g. 16312 -> 16.3k). The exact value could be shown as a tooltip or something.
> We do this in individual match pages, but not here. Could be added, but I figured it was more important to view the exact values since it matters more when you're for example sorting by that column and want to see minute differences.

DataTables supports independent display/sort values for this purpose. The exact value could be shown as a tooltip over the shortened number.

>> I don't expect to see the number of times they've played against their most played heroes. It's not that the data isn't useful, it's just in a strange place.
> the expected use case is for players to view their own profiles, and for that it seems logical to display data such as with/against for each hero they commonly play.

I understand that viewing your own profile is the "default" use case, I've no idea why I mentioned viewing other peoples' profiles at all really.

What I was really trying to say is that heroes I most commonly play with/against shouldn't be lumped into the same table as my most played heroes, these 3 datasets are not related.

YASP feels really fast this morning btw. Loving the new server upgrade.

**Comment by howardc93** [comment]
Status:
It allows you to see if matches are being skipped (for those who don't visit YASP regularly), or if you're viewing another profile that may not be 100% parsed, you might want to know at a glance which matches are parsed so you can view them (if you're trying to look at someone else's games).

Lobby type:
We can't differentiate, the RankedTeam MM lobby_type is from back in the day when the game had 5-stack team ranked MM, which has since been removed.

Abbreviations:
Yeah, we can probably add abbreviations to DataTables at some point.  @nickhh ?

Matchups:
All three data sets are related to heroes (played as, played with, played against).  In addition it might be interesting to note contrasts in the datasets (I win a lot playing X hero, but lose a lot when X hero is on my team, etc.).

If you have an idea for a different display format we can look into it.

Speedup:
Yes, we upgraded to the next tier of VM, and so far it seems to be running nicely.

**Comment by nickhh** [comment]
Abbreviations: I plan on adding additional tool-tips (see #453) but if you have any specific requests and/or wording suggestions, please let me know :)

**Comment by howardc93** [comment]
This would be updating DataTables to use a separate value for display and for sort in the match tables.  I think we can just call the format function directly from within DataTables, so no need to mess with css

**Comment by stuki** [comment]
Thing is with the colors that your red and green are kinda "neonish" for usage on the dark background compared to dotabuff for example: 
[light theme](https://dl.dropboxusercontent.com/u/854081/Pics/Screenshot%202015-07-01%2015.04.01.png)
[dark theme](https://dl.dropboxusercontent.com/u/854081/Pics/Screenshot%202015-07-01%2015.04.27.png)
[dotabuff](https://dl.dropboxusercontent.com/u/854081/Pics/Screenshot%202015-07-01%2015.05.06.png) 

That at least is the reason I chose to use the light theme

**Comment by LewisJackson1** [comment]
The green on white in the light theme from that screenshot is really hard to distinguish too.

**Comment by howardc93** [comment]
We can adjust the shade easily here:
https://github.com/yasp-dota/yasp/blob/master/public/css/yasp.css#L342

Find a hex code for a shade that works well on both themes?

**Comment by howardc93** [comment]
@stuki RE: light blue--are you talking about the font color or the adblock banner?  Light or dark?

**Comment by LewisJackson1** [comment]
Why not just the same as ".progress-bar-success" - #5cb85c?

**Comment by stuki** [comment]
@howardc93 The font color, clashes both against light and dark backgrounds

**Comment by howardc93** [comment]
@albertcui is planning to work on UI update this weekend, I'll tag him here so he remembers.  You could play with it yourself (use browser tools to manually find a shade that works well in both themes) and save us a bit of time :)

**Comment by dumasymptote** [comment]
#d9534f seems to work decently well in both themes for the link colors. It matches the color you are using for the loss text too.

**Comment by howardc93** [comment]
Link colors. . . so just drop it in an a {} in the CSS to apply it to all links?

**Comment by howardc93** [comment]
Update on light theme status:
```
db.players.count({last_visited:{$ne:null}, light_theme:true})
> 1520
```

This is out of nearly 20000 users, so light theme usage is approximately 8%.  I still want to get rid of it and have a unified design.  This also means we can get rid of the /preferences endpoint we currently use exclusively for this task.

**Comment by nickhh** [comment]
No. Stop trying to remove features. I use light theme. There's no reason to
kill something that already exists. Stop it.

**Comment by howardc93** [comment]
Well, I don't know if @albertcui 's new retheme will include two styles.  I am personally against it.

I am not planning to remove the existing option until the retheme occurs.

**Comment by howardc93** [comment]
It looks like Dotabuff uses the same shade of green for their links as they do for "won match".  We could do the same (basically the inverse of what @dumasymptote suggested).

**Comment by howardc93** [comment]
also @nickhh it's not about killing features that exist, but about maintainability.  Keeping two themes doubles our development/testing time because we have to write two sets of CSS and make sure changes look okay on both.

**Comment by vramana** [comment]
![p01hnzd](https://cloud.githubusercontent.com/assets/3112509/8807235/c8fa3548-2ff8-11e5-8724-7f9f091d089b.png)

I think pagination buttons should be some color other than green.

**Comment by howardc93** [comment]
I personally think they're fine but you can play around with the CSS if you
like.

**Comment by Iparke** [comment]
"Material design? Requires adding class names to all our templates :("

May I ask about the status of this concern? @albertcui 
I was talking to Howard about the UI. I have a personal vision for what I want the site to feel like and how the user experiences it, but at the same time, I do not want to go renegade on your plans.

I am drafting prototypes. I want to utilize Bootstrap, Node, and AngularJS. Data models would probably stick around for now, but eventually I would want to switch to D3.

**Comment by albertcui** [comment]
I'm currently making a migration to Material Design using Material Design
Lite. See here: http://dev.albertcui.com:8080/matches/1471454566/

But happy to work with you :) I'm not that far along, only have the first
two and a half match tabs done.

I'm not sure if we have any reason to use Angular. Seems like a lot of
overhead since we don't really need data binding...

D3 for graphs? Down for that! We haven't done that just because we don't
know any D3 haha.

**Comment by howardchung** [comment]
c3 is just a wrapper around d3.  So we're technically using it.

**Comment by Iparke** [comment]
Hey Albert, the connection to your website was refused, so I cannot see the page. Maybe post an image? :)

I can see your point about AngularJS. Admittedly, I have not looked into your guys' code base quite yet. I don't know if you guys are using MVC or whatnot. I don't know how you are getting data to the front end. I personally prefer AngularJS because it's easy for me to componentize and modularize everything. I like its routing. I like its built in directives. I am not versed enough to know the footprint of typical presentation Angular apps, but I would assume it's not going to be great, especially if we utilize a CDN to distribute it and especially because Angular apps are supposed to be more client intensive than server intensive. Admittedly, everything I just said is still managed without AngularJS.

I really like this Material Design Lite. I do not know it myself, but I am definitely willing to take a crack at it. *Everybody* uses Bootstrap :)

Sketches forthcoming, but my vision is to turn this into a fully responsive, scale-able, modular web application. I know how I use Dotabuff. Dotabuff in its current form is all about just shoving data out to the user without any experience considerations. They use a ton of bland tables with archaic dropdowns and pagination hailing from the ASP.net days. I want YASP to be fun. I want YASP to feed relevant information to the end user immediately and at a glance. I want end users to visualize their performance improvements. If I get my way, YASP will turn into a dashboard.

**Comment by albertcui** [comment]
Sounds great. Honestly the redesign is very much still tables :)

![screencapture-dev-albertcui-com-8080-matches-1471454566-1440631715947](https://cloud.githubusercontent.com/assets/3838552/9508973/f908f6d0-4c17-11e5-9075-b98a2b522d8a.png)

**Comment by Iparke** [comment]
Haha. I meant no offense. A lot of it will still be in tables. I want to implement functional, reactive tables that look good.

**Comment by howardchung** [comment]
I agree Angular is probably too much.  If we're going to use a JS framework I'd go with React.  View only, relatively lightweight.

We can define reusable components for tables.

We technically have an API for JS client to hit already, ?json=1.  However this returns ALL the data for a match/player, which is probably more than we need.

**Comment by Iparke** [comment]
For example, as a stats fan myself, I often scan my match pages for who has the highest GPM on the team. If I've lost, it certainly makes me feel better about myself if I have a decent GPM :P Users should be able to sort by clicking on the table headers.

**Comment by Iparke** [comment]
I will look into React. Like I said, there may not be a need at all for either. If this is a simple reskin with some added functionality, I'll just script my way.

**Comment by howardchung** [comment]
Currently we use Jade as a template to render our HTML server side.  Client side JS does last-mile things like heat maps, graphs, word clouds and sortable, pagable tables.

We're basically just avoiding the pagination issue by dumping all the data in one page and "paginating" client side.  It's not great when a user has thousands of matches.

**Comment by Iparke** [comment]
Thank you for the information. I feel like I'm here stirring the pot and without knowing everything, which is probably going to lead to bad communication somewhere down the line.

Regarding these "fixes": isn't something going to have to give sometime down the line? Paginating client side sounds as bad as it is considering some players have 10k+ hours of Dota 2. Is there a coherent way we can talk about infrastructure to make sure the front end can consume it properly?

**Comment by howardchung** [comment]
Yeah, it's not good.  The top players have just over 10,000 games at this point, and the number will just keep growing.

Considerations:
* We currently support client-side sorting on the full match table.  This lets users sort all their matches by fields like K/D/A/GPM.  We can add more, but ultimately we're limited by how many columns we can cram into one table.
* We could only have the server send a requested set of matches at once.  This means we have to have some kind of API for the client to hit.  In order to keep sorting alive, it also needs to be able to support sorting.
* The load on the server is just the same (since we have to grab all the matches and sort them, then slice out the page we want to send to the user), but we save bandwidth this way.
* If we're going to do client-side rendering with a server API, I want it to be done for the entire site.  I don't want to be doing client-side rendering for some bits and server-side on others.  Inconsistency is annoying.

**Comment by albertcui** [comment]
We had an API at one point, so the API route is very possible ;p

**Comment by Iparke** [comment]
I would love the API route. 

I'm working hard on these prototypes. I hope you guys will like them.

**Comment by howardchung** [comment]
API is still there (under routes/api.js).  It was designed to be a generic match API, but it may make more sense to simply use the players/account_id?json=1 (which does some post-processing of the raw match data).  

We'd need to add slicing and sorting though, which I don't think is currently supported.  IIRC, the keyword json is ignored in the query string, and everything else is treated as a select query--hero_id, game_mode, etc.

**Comment by Iparke** [comment]
A small preview of what I'm cooking (WIP). This is the dashboard, it greets the user (assuming the user is logged in already) as the default page.

![image](https://cloud.githubusercontent.com/assets/11266229/9513725/f1cb2a2a-4c47-11e5-858e-d6937f2e6a82.png)

I do not know your guys' end goal. My end goal is that YASP gets used. I don't use YASP. I want to use YASP. The only way YASP is not another Dotabuff (YADB) is if we make an effort to make YASP awesome. And the way we make it awesome is by giving users things to do while tracking their performance. I talk stats with my friends all the time. We sit around after a nice night of Doto and laugh at each other's stats. We compare each others stats. We have friendly competitions over our stats.

Here's the breakdown of the preview-:

**Profile picture and sleeve**: The user picks from preset profile/background combinations. We need to ask artists for their permission in order to use their art. We would offer a link back to their respective sites on the settings page during combination selection. Once we have the artists backing us, we get exposure.

**Titles**: Each user has a title. The title consists of:
1) The hero they spam.
2) An adjective (in this case 'accomplished') to describe their overall K/D/A (on all heroes).
3) The lane they most frequent.

**Endorsements**: These work exactly like LinkedIn endorsements. You can suggest tags for yourself, but only other users can vote to say if you have that quality. Available qualities are predetermined and all positive or neutral, no negative tags. The endorsements can range from player roles (carry, support) to strategies (ratter, pusher, nuker) to qualities (friendly, forgiving, leader). Some tags are only given by YASP administrators ("Pro" comes quickly to mind). You have to smile imagining how many people are going to search for Bulldog just to endorse him as a 'ratter.'

I hope you like it; it is far from finished. I have lots of more additions and plans, especially when it comes out to designating layout. I realize that I'm adding additional features, please let me know your thoughts.

**Comment by Iparke** [comment]
I should note, I'm doing all of this work in Inkscape -- it's not markup, it's not real.

**Comment by howardchung** [comment]
YASP started as a simple side project for fun by a couple of college students (@albertcui and myself).

Dotabuff already does a lot of fine things and provides a lot of statistical data to the community.  I respect them a lot for what they've achieved and for their open source contributions.  

However, we wanted to basically implement all the "Dotabuff Plus" features to give ourselves a free option (because I'm super cheap and hate spending money, I'd rather put in hundreds of hours of development time to build my own :P)

In my opinion, there's not really a reason to try to duplicate the things that Dotabuff is already offering for free (why reinvent the wheel?).  You're suggesting some interesting ideas and I think they might be good in order to make YASP a bit more distinctive.

* Art.  Someone will need to contact a lot of artists and secure permission.  It might be cool, but I don't want to be that person since it will be time-consuming and I'd rather be spending my time writing code.
* Titles.  Interesting idea.  Someone needs to come up with the buckets/determinations for the title, but I like this idea :)
* Endorsements.  @albertcui has talked about making YASP more "social-networky", including stuff like friends, activity feeds, and match comments.  This would simply be an extension of that.  I don't really mind it being there, but I'm not particularly interested in working on those features myself since I feel that they aren't the "core" of why people use YASP.
  * Caveat: Most of our player profiles are "incomplete", meaning that they're missing most matches.  This is due to the fact that we only retrieve a player's match history on login (to save disk space).  Currently we have around 8% of all the Dota 2 matches ever played.  This may preclude usage of YASP as a "social network".
  * Similarly we currently have no search functionality because of the fact that YASP is not a comprehensive database of all matches.  To do name searches we'll also need to bring in something like ElasticSearch, and we also need to deal with the fact that Steam personanames are mutable and change a LOT.  It's an annoying problem.

**Comment by Iparke** [comment]
"Similarly we currently have no search functionality because of the fact that YASP is not a comprehensive database of all matches. To do name searches we'll also need to bring in something like ElasticSearch, and we also need to deal with the fact that Steam personanames are mutable and change a LOT. It's an annoying problem."

Correct me if I am wrong, but doesn't Dotabuff only track matches after a user has joined Dotabuff? Couldn't YASP say that upfront during registration? 

I'm also curious why YASP tracks a user only after donation. 
I guess the reality of the situation is: YASP is more of a pet project than an actual 'product.' Which is not how I'm treating it right now haha. I don't want to make money off of it, but I want it to be used by the Dota base. We're not a business. How feasible is it to sustain a large user base (tracking everybody) with only ad revenue?

**Comment by howardchung** [comment]
Dotabuff indexes all matches.  Steam has a sequential API that can be used to obtain every match ever played.  We use the same one but skip matches without any registered users.

We don't say that we skip most matches "up front" since we generally put features/positives on the home page, and the "registration" is simply a Steam OAuth so we have no control over how that form looks.  We do mention it in the FAQ.

Yeah, it's a hobby project and so we sometimes push new features aggressively without a comprehensive review/test process.  That said, we still follow good engineering principles (code review, branches, CI, etc.) and avoid downtime/bugs.

We don't intend to make money from the project, since it's for fun rather than for a living, so we're looking to invest any surplus donations back into the project.

I personally would like to see it used by more people (just for the technical challenge of scaling and for the sense of accomplishment), and while our userbase is not huge, it's definitely grown a lot from our launch.

Some math regarding matches/parsing:

* Currently, there are approximately one million matches played per day.
* It's feasible to simply get the basic match data from the Steam API (what Dotabuff does) for all of these, at the cost of ~4GB (after compression) of database growth per day.
  * If we started adding all matches, we might as well go back and get every match ever played.  This would take roughly 2TB of storage, and would cost us $340 a month to keep on SSD (which we want to do for decent page load speeds).  This is a beyond our current budget.
* It is not feasible to do replay parsing on all these matches.  This would require a cluster of ~50 servers, along with 10,000 Steam accounts.  While our architecture (should) scale to this size, we don't have the budget for (at $40 a server/month, that's $2000 a month in server costs, not to mention the increased storage cost since a parsed match takes roughly 70kb compressed. 70kb*1million=70GB per day).  Not to mention Valve would probably notice and shut us down if we tried to make 10k accounts.
* Ad revenue covers only a fraction of our costs.  Increasing our scope to index/parse all matches would just increase our costs without increasing ad revenue (unless it convinced a lot more people to start using YASP)

**Comment by Iparke** [comment]
I suppose any technical debt YASP has accumulated by now runs more along the lines of "it would be nice to have" than "we need to dumpster YASP and start over." I think we can really work with the data and the system you two have already aggregated to create something exciting. I am not looking to take down Dotabuff. However, I would like YASP to take on its own niche identity.

What I am suggesting is breathing new life into YASP through user interactions. Somebody on Reddit said that they use Dotabuff to compare their own stats and they use YASP to compare their stats against their friends. If future drive goes towards this direction, YASP and Dotabuff can coexist and fill different roles in the community. I don't think YASP's current incarnation precludes social interaction.

Could I get feedback on this idea? I'm not going to fork your project. I would like to hear this is a direction that you guys want to take YASP. If it is a simple reskin you really want, hey, I'm still onboard, but if we do extend YASP to be more about the user than about the stats, then I want to hear that too. I am quite prepared to spend a lot of time on this project.

TL:DR I want to overhaul the definition of 'member' on YASP. I want commitment to the idea.

**Comment by albertcui** [comment]
*Profile picture and sleeve*: Looks good, but don't know if it really adds anything besides looking good.

*Titles*: Love it

*Endorsements*: Great as well

As @howardchung has said, this falls in line with what I wanted to do eventually for YASP. Some friend and comment system, and a newsfeed where you can see all the recent games you and your friends have played, with some quick overview/unique stats from each game.

The end goal is to give as many users as possible a free, fun, useful product in a cost effective manner.

**Comment by albertcui** [comment]
You should also note, we've seen an overwhelming preference for dark themes, so you should keep that in mind when designing the UI. We used to have both a white and dark theme, but it got to be a lot of work to keep maintaining it.

**Comment by Iparke** [comment]
Thanks for the comments @albertcui .

The **profile picture and sleeve* is there just for aesthetics. Gives the user something to customize, mostly there to fill up space. Maybe gives the user a feeling that YASP is home.

By the way, I've been doing a good amount of research on Material Design Lite.

Pros:
* It's fresh. It's new. It's stable.
* It's built to be mobile first and therefore responsive.
* Awesome feeling premade forms and inputs. (All of them come with mouse click outward sploosh support)

Cons:
* I've been reading up on the style guides. Greece's economic director has more options than Material Design users. For example, how to draw a human shape is explicitly defined in the Docs. Gotta base your images on three primitive geometry shapes *or else*. Gotta use this color palette *or else*.
* I think high contrast, saturated and strange vibrant colors are gross. (We don't have to use them, but Google might have us offed)
* The grid system is too primitive. Not as developed as Bootstrap. MDL has three different grids: tiny tablet, tablet sized tablet, and desktop.
* IF we want strange colors and icons, Bootstrap has a Material Design port. Look and feel of Material Design, but feature set of Bootstrap.
* At the time of this writing, no functional table support from MDL. Bootstrap has support for table operations. The MDL tables are extremely pleasing to the eye though..

I am leaning towards Bootstrap still. I can be convinced. Thoughts?

**Comment by howardchung** [comment]
@albertcui should probably weigh in on the direction we want to go.  I would personally like to spend most of my time on stats and visualizations, but I'm not opposed to adding social features.

If you can implement them yourself, awesome!  Otherwise it will depend on how much @albertcui and @nickhh want to do it.

**Comment by Iparke** [comment]
I can do it. I'm not a web developer. I am a software engineer with hardware background.

**Comment by albertcui** [comment]
Re Material Design:

* The style guides are just guides. Google itself doesn't even follow them half the time. I think in general MDL looks better than Bootstrap, which means less custom styling needed in the end.
* I was actually pretty happy with the color pallet I posted haha. What do you think?
* The grid seems as powerful as Bootstrap. One thing I noticed is that it's missing column offsets.
* What do you mean by table support?

I think in general, Bootstrap's API is _much_ better than MDL though. I just much prefer MDL's look over Bootstrap.

**Comment by howardchung** [comment]
A concern with social features:

These tend to only be useful if we have a large userbase (look at Google+).  With something like 6000 weekly active users I'm concerned that we won't have enough users to drive them.

**Comment by howardchung** [comment]
@albertcui we should probably be using bootstrap with a MDL theme then lol.  You've already invested quite a bit of time into retheme though

**Comment by albertcui** [comment]
Well I think also:
* Social features can help us grow the user base
* My guess is YASP users have friends who are YASP users

**Comment by Iparke** [comment]
@albertcui: You are right. Sorry. I was reading comment threads on an older iteration. The tables do have sorting on them. And the grid system seems to have beefed up. I'm off the MDL hate train. Let's do it in MDL. 

@howardchung: What Albert just wrote.

**Comment by Iparke** [comment]
@albertcui: And concerning the color scheme.. Let's decide together :)

**Comment by Iparke** [comment]
Seriously though. If we follow through with even half the ideas I have in my head right now, there's no reason for anybody to use Dotabuff. Dotabuff Truesight for free with a great UI that supports bragging to my friends? We might kill Dotabuff. They'd have to compete.

**Comment by stefan-j** [comment]
Any reason for leaning toward mdl? I think Bootstrap is the safer (better) option. 
If you enjoy the look of mdl you could always consider a mdl bootstrap template. Bootstrap has has a good stable API.

**Comment by howardchung** [comment]
I would also prefer to stick with Bootstrap since it's heavily tested and stable.  It would save me the time of having to learn all the right class names/syntax for MDL.

However @albertcui has already invested significant time into his retheme, so if he wants to go through with it I will accept it.

**Comment by albertcui** [comment]
MDL is currently version v1. Have there been comments that the API will change in the future?

I honestly like MDL's navigational components much better than Bootstrap: nav bars, navigation drawer, footer, tabs.

MDL is less comprehensive, but that also means writing less CSS to override the default style components when we want to change something. IMO it's more freedom.

**Comment by Iparke** [comment]
^ IMO: As long as my commits make it into the project, custom CSS and going beyond default styles is going to be a thing. I'm not sure if that makes it a compelling argument for MDL if all I am going to do is customize it to the nth degree. Bootstrap has far more support in a larger following, in case we do hit walls -- and we will. I'm still on board with MDL just because you are enthusiastic about it and I don't mind learning. Bootstrap is a compelling option.

**Comment by vramana** [comment]
Hi folks,

For me personally, Dotabuff is way ahead of YASP in UX because of its clean and minimal UI. I think we should concentrate on our UX for now. YASP is provides a lot of statistics, people are missing a clean way of consuming them. 

I think we can work on the current design and make it a much better UI rather than complete rewrite. Part of the problem with the current UI is we use Bootstrap elements directly without theming them accordingly. Just by nuking a whitespace alone and readjusting things, I feel the YASP UX can be improved a lot. I am specifically talking about tables. They just consume too much space for the amount of information they show. They can be made significantly compact.
 
But if you decide rewriting the way to go. Then these are somethings I would like to see as part of next design.
1. Flat & Compact look
2. Single Page Application 
3. More Animations (eg. Page transistions, Filter animations) 

I think React along with React-Router works nicely for this description.  I think we can simply limit ourselves using just grid system in Bootstrap and rest we can use from other libraries.

**Comment by Iparke** [comment]
@vramana 

1. Flat & Compact look
I am also a huge fan of flat design. How flat are you thinking? Is this flat enough? 
![image](https://cloud.githubusercontent.com/assets/11266229/9553311/ffbbcc62-4d72-11e5-910f-b7db8f87d719.png)

2. Single Page Application 
I want to turn YASP into a Dashboard :D

3. More Animations (eg. Page transistions, Filter animations)
In terms of page switching, I really like Microsoft's Azure. https://youtu.be/aNYtpb3-Dxc?list=PLLasX02E8BPBZjy3YLHrN0LC9QO2CSa6H&t=11
I have a pretty awesome idea for how to implement filters, but I'm saving it for later :)

**Comment by albertcui** [comment]
What's the advantage of a single page application? Maybe for things like Player pages, but I'm not sure I see the advantage for match pages.

**Comment by howardchung** [comment]
concerns with SPA design (assuming we turn match/player pages into SPAs with the tabs just swapping in different data without a reload):

* We lose metrics that tell us where people are visiting (since Google Analytics will treat them all on one page regardless of what tab they're on).
* It becomes harder to link people to a particular tab.  People like sharing their word clouds, etc.  Perhaps we could solve this by tacking on the current tab name as a hash and having JS render the correct section based on it.
* Slower initial load?  We have to get/render the entire match/player at initial load.

**Comment by albertcui** [comment]
* I was wondering about Google Analytics, too, but looks like [GA supports it](https://developers.google.com/analytics/devguides/collection/analyticsjs/single-page-applications)
* I'm sure there's a solution for this as well. The JS framework just needs to update the URL.
* Slower initial load + increased bandwidth unless we do the API!

**Comment by howardchung** [comment]
API still means we have to download the entire match's data.  Right now we only send over the wire the HTML necessary for that page.

**Comment by albertcui** [comment]
Not really, the API could be more specific and only send the necessary information.

**Comment by howardchung** [comment]
true, but then we have to implement all that selectivity from scratch.

**Comment by Iparke** [comment]
We can still use restful routes on SPA. And we don't need all the data upfront, we can make AJAX calls as needed.

**Comment by howardchung** [comment]
It can certainly be done.  The question we need to ask is: What benefits do we get out of it in exchange for the development time we put in?

I also don't want to do "hybrid rendering" with some on the server side and some of the client side.   If we swapped to SPA/client side rendering at all, I'd want it to be integrated to the point where we can drop Jade completely and just serve up HTML/JS that renders the page by hitting API endpoints.

**Comment by howardchung** [comment]
Also I think implementing #405 would help our tables look better and be more visually striking :)

**Comment by albertcui** [comment]
Isn't hybrid rendering always a requirement? Things like headers and footers are probably same across the application.

**Comment by howardchung** [comment]
Well, we could either have a single HTML page that we serve and the JS within decides what to render.

Or we could use HTML5 imports: http://www.html5rocks.com/en/tutorials/webcomponents/imports/

**Comment by Iparke** [comment]
"I also don't want to do "hybrid rendering" with some on the server side and some of the client side. If we swapped to SPA/client side rendering at all, I'd want it to be integrated to the point where we can drop Jade completely and just serve up HTML/JS that renders the page by hitting API endpoints."

Please. I do not like Jade.
Let me go Angular, give me an API, and I'll hash out this project in no time. :)

"Well, we could either have a single HTML page that we serve and the JS within decides what to render."
Angular directives can be just as good as HTML imports.. :)

**Comment by howardchung** [comment]
I want to hear this idea about filters.  Traditional select boxes are boring :(

I think React better suits our purposes and is more future-proof.  Angular's in a weird place right now between the 1 and 2 split, and is more than we really need.  Our data flow is unidirectional.

An API of sorts already exists.  If you want to play around a bit you can try hitting a match url + ?json=1 to get back the entire match JSON data.

**Comment by vramana** [comment]
@howardchung  All the routing can be handled with React-Router so that each page gets its own URL for our SPA. Rather me telling please watch this talk https://www.youtube.com/watch?v=Q6Kczrgw6ic I think this will address all our concerns.  I didn't understand your stance on server-side rendering. What is the issue here?

**Comment by howardchung** [comment]
The question remains: what benefits do we get out of spending time implementing this?

So far I can only think of one:
* Ability to do client side parsing with JS (for example, with a user's own private lobby replay) and display match report to user without any need to interact with the server.

Another concern is how to have constants defined (id->friendly name mappings).  With the server we can do those lookups at render time but with an API and client-side render we need to figure out a different way of passing those values to the client.

**Comment by Iparke** [comment]
"The question remains: what benefits do we get out of spending time implementing this?"

API:
- Code reuse!
- Code structure.
- Opens potential partnerships with external third parties who can consume our API.
- If additional features need to be implemented, existing APIs can be modified with quick turnarounds.
- APIs allow different presentation layers to consume the same data, but in a way they want to consume it.

SPA:
Comes down to user experience.
Shameless copy and paste from (http://designmodo.com/ux-single-page-websites/):
- One-page navigation is more straightforward; there is no way to get lost and everything is just a couple of scrolls away.
- The page is more focused and clear, which makes it easier to communicate the main message of the website.
- Single-page websites work smoothly on almost all devices, since the navigation is mainly though scrolling or swiping instead of clicking or tapping.

Ultimately, here's the scenario as I see it. Traditional and SPA sites, when built optimally, pretty much the same deal. There is no *real* reason we should choose SPA over traditional or traditional over SPA (maybe SEO). It does boil down to personal preference. I prefer SPA because of user experience. The way I see it, if UX is the big complaint for YASP and we are doing an overhaul, we should go big or go home. I can see the argument that all we need to do here is a reskin, but I don't want to do a reskin. A reskin makes things prettier, sure, but it's just putting lipstick on a pig. If we're going to consider UX and put it as a priority, let's go all out.

**Comment by howardchung** [comment]
The thing about YASP is that I'm not sure migrating to SPA would change the UX all that much.  Users primarily care about two sets of pages--matches and players.

We obviously have too much data per match/player to just spit it all out at once--so the user is going to have to click tabs or navigation of some kind.  That's not really different UX from how it is now.

Can you point out a specific example of a  UX element that would be improved by switching to SPA?

**Comment by albertcui** [comment]
I think SPA with an API will load faster, filtering of matches (adv query), pagination of matches, we could even side by side load player pages for comparison which would be easier with SPA...

SPA makes sense.

Re Bootstrap vs MDL: we can use whatever you guys want. I will continue my MDL retheme because I need something to do and it can be used as a placeholder while we get the "final" theme done.

**Comment by Iparke** [comment]
I'm okay with that. Sorry I haven't been replying as much lately. Life outside of work has been a little busier these days.

People should stop getting married.

**Comment by howardchung** [comment]
I'm not sure it will be much faster.  The heavy lifting (filtering/querying) still has to be done on the server.  The only time we save is on rendering, which only really matters for large result sets (such as all matches for a player)

**Comment by albertcui** [comment]
For the user it will *appear* faster since the UI remains. Instead of
looking at a blank, white browser loading screen, we can throw up a loading
icon, etc.

**Comment by howardchung** [comment]
So do this for players only?  Matches load pretty fast already.  Cached
players are super fast as well, the only slow pages are uncached players,
which can take a few seconds (scaling with number of matches)

**Comment by howardchung** [comment]
http://blog.getbootstrap.com/2015/08/19/bootstrap-4-alpha/

**Comment by howardchung** [comment]
Closing this issue due to it getting really big.  Please open new issues for specific UI fixes.
